### PR TITLE
shared-mime-info: clean up cache on deactivate + auto-update watcher/daemon

### PR DESCRIPTION
**Pull request by RJVB**
#### Description
Clean the mime cache after deactivating the port, as requested in https://trac.macports.org/ticket/45396 .

_Note that in my testing and experience the `post-activate` step is only performed after an install. So this change means that the mime cache is removed after a `port deactivate shared-mime-info` so also after a `port activate shared-mime-info` to activate a different version - in which case a new cache is not built automatically.

I see no way to warn the user of this fact other than after an install/upgrade, which is why I am not in favour of this new behaviour personally._

This PR also proposes to replace the post-activate call to `update-mime-database` with a launchd daemon watching `$prefix/share/mime/packages`. To make this change as transparent as possible the daemon's plist is loaded in the `post-activate` and unloaded in the `pre-deactivate`; the user is informed of this via `port info` and `port notes` (which also contains instructions related to the re-activation issue mentioned above).

If this change is accepted all dependent Portfiles could lose their post-activate call to `update-mime-database`.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
Mac OS X 10.9.5

###### Verification 
Have you

- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ X] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [ X] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by mascguy** [comment]
Hey Rene, I completely forgot about this PR! But finally trying to test with it locally.

I did encounter one problem, which is that the launchd script appears to be missing: `org.macports.shared-mime-info-updater.plist`. I could probably cobble one together easily enough, as presumably (?) it simply runs:

- `env XDG_DATA_DIRS=${prefix}/share ${prefix}/bin/update-mime-database -V ${prefix}/share/mime`

But was hoping you still have that file in your repo? If so, would you mind adding that?

**Comment by RJVB** [comment]
I did have the file and added it to my fork via the web interface but it doesn't seem the PR notited that change...

**Comment by mascguy** [comment]
> I did have the file and added it to my fork via the web interface
> (note to self: this works if you just watch to which fork you're adding)

Thanks Rene! I'll test this later today.

**Comment by mascguy** [comment]
> > I did have the file and added it to my fork via the web interface
> > (note to self: this works if you just watch to which fork you're adding)
> 
> Thanks Rene! I'll test this later today.

Looks good overall. One thought though: As you know, it's preferable to utilize MacPorts' `startupitem` functionality, rather than directly installing a Launch Agent/Daemon.

Is there a specific reason you didn't follow that approach?

File: devel/shared-mime-info/Portfile
Comment (mascguy) [inline devel/shared-mime-info/Portfile:1]: You need to rev-bump this, of course...

**Comment by RJVB** [comment]
>Is there a specific reason you didn't follow that approach?

TBH I didn't know it was preferable, am not particularly fond of black boxes ;) but apart from that I see no reason to avoid it if it can do the same thing. Note that having the mime info updating done correctly means the agent has to be loaded automatically at the activation stage and not depend on the user doing it. If the startupitem approach has a way of doing that without calling an external application in the post-activation phase (and reliably, not just in the 1st one after an install) then that would in fact be preferable.

**Comment by mascguy** [comment]
> TBH I didn't know it was preferable, am not particularly fond of black boxes ;) but apart from that I see no reason to avoid it if it can do the same thing. Note that having the mime info updating done correctly means the agent has to be loaded automatically at the activation stage and not depend on the user doing it. If the startupitem approach has a way of doing that without calling an external application in the post-activation phase (and reliably, not just in the 1st one after an install) then that would in fact be preferable.

While users can technically disable automatic loading of startup items - by setting `startupitem_autostart no` in `${prefix}/etc/macports/macports.conf`, very few users do. And we can add notes, to display a message to reiterate the importance of enabling the item for this port.

So I'd like to suggest we use MacPorts' startupitem functionality, rather than installing the agent directly.

**Comment by RJVB** [comment]
>While users can technically disable automatic loading of startup items - by setting `startupitem_autostart no` in `${prefix}/etc/macports/macports.conf`, very few users do. And we 

I updated my DBus port not long ago, and IIRC I did have to activate it manually. For full disclosure I have to admit I upgraded it from a very old version that didn't use startupitem functionality.

**Comment by mascguy** [comment]
> > While users can technically disable automatic loading of startup items - by setting `startupitem_autostart no` in `${prefix}/etc/macports/macports.conf`, very few users do.

> I updated my DBus port not long ago, and IIRC I did have to activate it manually. For full disclosure I have to admit I upgraded it from a very old version that didn't use startupitem functionality.

Yep, that's the usual case. So are you willing to switch to using `startupitem`?

**Comment by RJVB** [comment]
>Yep, that's the usual case. So are you willing to switch to using `startupitem`?

If we're clear that the usual case isn't good enough here and that it can be made automatic... and maybe if you show me how ;) (I'm guess it'd take you less than than me...)

**Comment by mascguy** [comment]
> > Yep, that's the usual case. So are you willing to switch to using `startupitem`?

> If we're clear that the usual case isn't good enough here and that it can be made automatic... and maybe if you show me how ;) (I'm guess it'd take you less than than me...)

Sure, I'm happy to make the modifications. But you'll need to enable `Allow edits and access to secrets by maintainers` for this PR, otherwise I can't push to your branch.

**Comment by RJVB** [comment]
So that's what that's for :) I thought you'd just copy the snippet here, but this is more efficient :)

**Comment by mascguy** [comment]
> So that's what that's for :) I thought you'd just copy the snippet here, but this is more efficient :)

Yep! Just pushed the changes, after testing and validating locally.

Let me know what you think!

**Comment by mascguy** [comment]
Actually, reviewing the port notes, we have some redundancy:

```
--->  Some of the ports you installed have notes:
    shared-mime-info has the following notes:
  Please check if the auto-updater is still active after reactivating (another) version of) this port: 
   `sudo launchctl list | fgrep shared-mime-info-updater` 
   If the updater is not active, load it manually with 
   `sudo port load shared-mime-info`
  
  A startup item has been generated that will start shared-mime-info with launchd, and will be enabled automatically on activation.
  Execute the following command to manually _disable_ it:
  
      sudo port unload shared-mime-info
```

Since MacPorts base already provides default notes for the startupitem, that might be sufficient. What do you think?

**Review by mascguy — APPROVED** [review]

**Comment by mascguy** [comment]
Rene, let's merge. I'm happy to refine the notes later, if you have any strong preferences.

Otherwise, thank you so much for taking this on. Great stuff, and much appreciated!

**Comment by RJVB** [comment]
Well, thanks for the assistance! You basically caught me offline due to time differences so sorry for not reacting in more timely fashion.

FWIW, how reliable is the startupitem activation after activation? On my end the post-activation phase only seems to trigger after a port install, not when activating another already installed version. Never yet got around to looking into the reasons for that.

**Comment by mascguy** [comment]
> Well, thanks for the assistance! You basically caught me offline due to time differences so sorry for not reacting in more timely fashion.

No worries, glad to help. You did the hard work, so the simplification was easy!

> FWIW, how reliable is the startupitem activation after activation? On my end the post-activation phase only seems to trigger after a port install, not when activating another already installed version. Never yet got around to looking into the reasons for that.

It looks like `post-activate` doesn't run after a deactivation/activation, though perhaps there's a reason for that behavior. (Or perhaps it's a known issue, that simply hasn't been addressed yet. Would have to search the open Trac tickets to say for sure.)

Thankfully this isn't an issue for Darwin/macOS though, as the startup item is consistently re-added/started on activation. So no worries there at least.

**Comment by RJVB** [comment]
>Thankfully this isn't an issue for Darwin/macOS though

Are you suggesting the issue isn't moot elsewhere?

I *have* been tinkering with/on a reverse-engineered launchctl/launchd clone for *NIX with someone in the beginning of the year, but as so often I got completely side-tracked and the other person probably lost interest by now :-/

**Pull request merged**